src/util/fs.ts

```diff
@@ -0,0 +1,269 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// Like node fs/promises, only we maintain a budget of open files to prevent
+// running out of file descriptors.
+
+import type * as fsTypes from 'fs';
+import * as fs from 'fs/promises';
+import {
+  createReadStream as rawCreateReadStream,
+  createWriteStream as rawCreateWriteStream,
+} from 'fs';
+import {Deferred} from './deferred.js';
+export {constants} from 'fs';
+
+declare global {
+  interface SymbolConstructor {
+    readonly dispose: unique symbol;
+  }
+}
+
+// The interface for https://github.com/tc39/proposal-explicit-resource-management
+// Once we're on TS 5.2 we can write a lot of this with `using` syntax.
+interface Disposable {
+  [Symbol.dispose](): void;
+}
+
+// Polyfill Symbol.dispose.
+// eslint-disable-next-line
+(Symbol as any).dispose ??= Symbol('Symbol.dispose');
+
+export class Semaphore {
+  #remaining: number;
+  readonly #waiting: Deferred<void>[] = [];
+
+  constructor(numSlots: number) {
+    if (numSlots <= 0) {
+      throw new Error(`numSlots must be positive, got ${numSlots}`);
+    }
+    this.#remaining = numSlots;
+  }
+
+  async reserve(): Promise<Disposable> {
+    while (this.#remaining === 0) {
+      const deferred = new Deferred<void>();
+      this.#waiting.push(deferred);
+      await deferred.promise;
+    }
+    this.#remaining--;
+    let disposed = false;
+    return {
+      [Symbol.dispose]: () => {
+        if (disposed) {
+          return;
+        }
+        disposed = true;
+        this.#remaining++;
+        if (this.#waiting.length > 0) {
+          this.#waiting.pop()?.resolve();
```

Comment:
Any risk that if this is treated as a stack in a resource constrained environment then it could starve any waiting promises at the base of the stack? Vs modeling this as a queue and giving preference to the longest waiting semaphore. 

Maybe this is not an issue in practice.

Reply:
This is definitely an unfair queue, but my thinking is that we're optimizing for throughput not latency, and so the only important thing is that we're making forward progress with minimal overhead, and pushing and popping off the end of an array is the fastest thing we can do here